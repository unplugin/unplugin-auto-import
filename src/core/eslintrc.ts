import type { Import } from 'unimport'
import type { ESLintrc } from '../types'

export function generateESLintConfigs(
  imports: Import[],
  eslintrc: ESLintrc,
) {
  const eslintConfigs: any = { globals: {} }

  imports
    .map(i => i.as ?? i.name)
    .filter(Boolean)
    .sort()
    .forEach((name) => {
      eslintConfigs.globals[name] = eslintrc.globalsPropValue
    })
  const jsonBody = JSON.stringify(eslintConfigs, null, 2)
  return jsonBody
}
