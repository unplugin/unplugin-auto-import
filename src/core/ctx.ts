import type { Import, InlinePreset } from 'unimport'
import type { BiomeJS, ESLintrc, ImportExtended, Options } from '../types'
import { existsSync, promises as fs } from 'node:fs'
import { dirname, isAbsolute, relative, resolve } from 'node:path'
import process from 'node:process'
import { slash, throttle, toArray } from '@antfu/utils'
import { applyEdits, modify, parse as parseJsonc } from 'jsonc-parser'
import { isPackageExists } from 'local-pkg'
import MagicString from 'magic-string'
import { createUnimport, resolvePreset } from 'unimport'
import { createFilter } from 'unplugin-utils'
import { presets } from '../presets'
import { generateESLintConfigs } from './eslintrc'
import { resolversAddon } from './resolvers'

export const INCLUDE_RE_LIST = [/\.[jt]sx?$/, /\.astro$/, /\.vue$/, /\.vue\?vue/, /\.vue\.[tj]sx?\?vue/, /\.svelte$/]
export const EXCLUDE_RE_LIST = [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/]

export function createContext(options: Options = {}, root = process.cwd()) {
  root = slash(root)

  const {
    dts: preferDTS = isPackageExists('typescript'),
    dirsScanOptions,
    dirs,
    vueDirectives,
    vueTemplate,
  } = options

  const eslintrc: ESLintrc = options.eslintrc || {}
  eslintrc.enabled = eslintrc.enabled === undefined ? false : eslintrc.enabled
  eslintrc.filepath = eslintrc.filepath || './.eslintrc-auto-import.json'
  eslintrc.globalsPropValue = eslintrc.globalsPropValue === undefined ? true : eslintrc.globalsPropValue

  const biomejs: BiomeJS = options.biomejs || {}
  biomejs.enabled = biomejs.enabled === true

  const dumpUnimportItems = options.dumpUnimportItems === true
    ? './.unimport-items.json'
    : options.dumpUnimportItems ?? false

  const resolvers = options.resolvers ? [options.resolvers].flat(2) : []

  // When "options.injectAtEnd" is undefined or true, it's true.
  const injectAtEnd = options.injectAtEnd !== false

  const unimport = createUnimport({
    imports: [],
    presets: options.packagePresets?.map(p => typeof p === 'string' ? { package: p } : p) ?? [],
    dirsScanOptions: {
      ...dirsScanOptions,
      cwd: root,
    },
    dirs,
    injectAtEnd,
    parser: options.parser,
    addons: {
      addons: [
        resolversAddon(resolvers),
        {
          name: 'unplugin-auto-import:dts',
          declaration(dts) {
            return `${`
/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// noinspection JSUnusedGlobalSymbols
// Generated by unplugin-auto-import
${dts}`.trim()}\n`
          },
        },
      ],
      vueDirectives,
      vueTemplate,
    },
  })

  const importsPromise = flattenImports(options.imports)
    .then((imports) => {
      if (!imports.length && !resolvers.length && !dirs?.length)
        console.warn('[auto-import] plugin installed but no imports has defined, see https://github.com/antfu/unplugin-auto-import#configurations for configurations')

      const compare = (left: string | undefined, right: NonNullable<(Options['ignore'] | Options['ignoreDts'])>[number]) => {
        return right instanceof RegExp
          ? right.test(left!)
          : right === left
      }

      options.ignore?.forEach((name) => {
        const i = imports.find(i => compare(i.as, name))
        if (i)
          i.disabled = true
      })

      options.ignoreDts?.forEach((name) => {
        const i = imports.find(i => compare(i.as, name))
        if (i)
          i.dtsDisabled = true
      })

      return unimport.getInternalContext().replaceImports(imports)
    })

  const filter = createFilter(
    options.include || INCLUDE_RE_LIST,
    options.exclude || EXCLUDE_RE_LIST,
  )
  const dts = preferDTS === false
    ? false
    : preferDTS === true
      ? resolve(root, 'auto-imports.d.ts')
      : resolve(root, preferDTS)

  const multilineCommentsRE = /\/\*.*?\*\//gs
  const singlelineCommentsRE = /\/\/.*$/gm
  const dtsReg = /declare\s+global\s*\{(.*?)[\n\r]\}/s
  const componentCustomPropertiesReg = /interface\s+ComponentCustomProperties\s*\{(.*?)[\n\r]\}/gs
  function parseDTS(dts: string) {
    dts = dts
      .replace(multilineCommentsRE, '')
      .replace(singlelineCommentsRE, '')

    const code = dts.match(dtsReg)?.[0]
    if (!code)
      return

    // eslint-disable-next-line regexp/no-super-linear-backtracking, regexp/no-misleading-capturing-group
    return Object.fromEntries(Array.from(code.matchAll(/['"]?(const\s*[^\s'"]+)['"]?\s*:\s*(.+?)[,;\r\n]/g)).map(i => [i[1], i[2]]))
  }

  async function generateDTS(file: string) {
    await importsPromise
    const dir = dirname(file)
    const originalContent = existsSync(file) ? await fs.readFile(file, 'utf-8') : ''
    const originalDTS = parseDTS(originalContent)
    let currentContent = await unimport.generateTypeDeclarations({
      resolvePath: (i) => {
        if (i.from.startsWith('.') || isAbsolute(i.from)) {
          const related = slash(relative(dir, i.from).replace(/\.ts(x)?$/, ''))
          return !related.startsWith('.')
            ? `./${related}`
            : related
        }
        return i.from
      },
    })
    const currentDTS = parseDTS(currentContent)!
    if (options.vueTemplate) {
      currentContent = currentContent.replace(
        componentCustomPropertiesReg,
        $1 => `interface GlobalComponents {}\n  ${$1}`,
      )
    }
    if (originalDTS) {
      Object.keys(currentDTS).forEach((key) => {
        originalDTS[key] = currentDTS[key]
      })
      const dtsList = Object.keys(originalDTS).sort().map(k => `  ${k}: ${originalDTS[k]}`)
      return currentContent.replace(dtsReg, () => `declare global {\n${dtsList.join('\n')}\n}`)
    }

    return currentContent
  }

  async function generateESLint() {
    return generateESLintConfigs(await unimport.getImports(), eslintrc)
  }

  async function updateBiomeConfig(dtsPath: string) {
    const biomeConfigPaths = [
      resolve(root, 'biome.json'),
      resolve(root, 'biome.jsonc'),
    ]

    for (const configPath of biomeConfigPaths) {
      if (existsSync(configPath)) {
        try {
          const content = await fs.readFile(configPath, 'utf-8')
          const config = parseJsonc(content)

          // Ensure files.includes exists
          if (!config.files) {
            config.files = {}
          }
          if (!config.files.includes) {
            config.files.includes = []
          }

          // Get just the filename of the dts file
          const dtsFilename = dtsPath.split('/').pop()
          const includesPattern = `!**/${dtsFilename}`

          // Check if already ignored
          if (!config.files.includes.includes(includesPattern)) {
            // Detect indentation from the original content
            const indentMatch = content.match(/^(\t| +)/m)
            const indent = indentMatch ? indentMatch[1] : '  '
            const tabSize = indent === '\t' ? 1 : indent.length

            // Use modify to preserve formatting and comments
            const edits = modify(content, ['files', 'includes', config.files.includes.length], includesPattern, {
              formattingOptions: {
                insertSpaces: indent !== '\t',
                tabSize,
              },
            })

            const newContent = applyEdits(content, edits)
            await fs.writeFile(configPath, newContent, 'utf-8')
          }

          break // Only update the first found config
        }
        catch (e) {
          console.error(`[unplugin-auto-import] Failed to update ${configPath}:`, e)
        }
      }
    }
  }

  const writeConfigFilesThrottled = throttle(500, writeConfigFiles, { noLeading: false })

  async function writeFile(filePath: string, content = '') {
    await fs.mkdir(dirname(filePath), { recursive: true })
    return await fs.writeFile(filePath, content, 'utf-8')
  }

  let lastDTS: string | undefined
  let lastESLint: string | undefined
  let lastUnimportItems: string | undefined

  async function writeConfigFiles() {
    const promises: any[] = []
    if (dts) {
      promises.push(
        generateDTS(dts).then(async (content) => {
          if (content !== lastDTS) {
            lastDTS = content
            await writeFile(dts, content)

            // Update biome config if enabled
            if (biomejs.enabled) {
              await updateBiomeConfig(dts)
            }
          }
        }),
      )
    }
    if (eslintrc.enabled && eslintrc.filepath) {
      const filepath = eslintrc.filepath
      promises.push(
        generateESLint().then(async (content) => {
          if (filepath.endsWith('.cjs'))
            content = `module.exports = ${content}`
          else if (filepath.endsWith('.mjs') || filepath.endsWith('.js'))
            content = `export default ${content}`

          content = `${content}\n`
          if (content.trim() !== lastESLint?.trim()) {
            lastESLint = content
            return writeFile(eslintrc.filepath!, content)
          }
        }),
      )
    }

    if (dumpUnimportItems) {
      promises.push(
        unimport.getImports().then((items) => {
          if (!dumpUnimportItems)
            return
          const content = JSON.stringify(items, null, 2)
          if (content !== lastUnimportItems) {
            lastUnimportItems = content
            return writeFile(dumpUnimportItems, content)
          }
        }),
      )
    }

    return Promise.all(promises)
  }

  async function scanDirs() {
    await unimport.modifyDynamicImports(async (imports) => {
      const exports_ = await unimport.scanImportsFromDir() as ImportExtended[]
      exports_.forEach(i => i.__source = 'dir')
      return modifyDefaultExportsAlias([
        ...imports.filter((i: ImportExtended) => i.__source !== 'dir'),
        ...exports_,
      ], options)
    })

    writeConfigFilesThrottled()
  }

  async function transform(code: string, id: string) {
    await importsPromise

    const s = new MagicString(code)

    await unimport.injectImports(s, id)

    if (!s.hasChanged())
      return

    writeConfigFilesThrottled()

    return {
      code: s.toString(),
      map: s.generateMap({ source: id, includeContent: true, hires: true }),
    }
  }

  return {
    root,
    dirs,
    filter,
    scanDirs,
    writeConfigFiles,
    writeConfigFilesThrottled: writeConfigFilesThrottled as () => void,
    transform,
    generateDTS,
    generateESLint,
    unimport,
  }
}

export async function flattenImports(map: Options['imports']): Promise<Import[]> {
  const promises = await Promise.all(toArray(map)
    .map(async (definition) => {
      if (typeof definition === 'string') {
        if (!presets[definition])
          throw new Error(`[auto-import] preset ${definition} not found`)
        const preset = presets[definition]
        definition = typeof preset === 'function' ? preset() : preset
      }

      if ('from' in definition && 'imports' in definition) {
        return await resolvePreset(definition as InlinePreset)
      }
      else {
        const resolved: Import[] = []
        for (const mod of Object.keys(definition)) {
          for (const id of definition[mod]) {
            const meta = {
              from: mod,
            } as Import
            if (Array.isArray(id)) {
              meta.name = id[0]
              meta.as = id[1]
            }
            else {
              meta.name = id
              meta.as = id
            }
            resolved.push(meta)
          }
        }
        return resolved
      }
    }))

  return promises.flat()
}

function modifyDefaultExportsAlias(imports: ImportExtended[], options: Options): Import[] {
  if (options.defaultExportByFilename) {
    imports.forEach((i) => {
      if (i.name === 'default')
        i.as = i.from.split('/').pop()?.split('.')?.shift() ?? i.as
    })
  }

  return imports as Import[]
}
