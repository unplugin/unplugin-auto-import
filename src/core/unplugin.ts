import type { Options } from '../types'
import path from 'node:path'
import { slash } from '@antfu/utils'
import { isPackageExists } from 'local-pkg'
import pm from 'picomatch'
import { createUnplugin } from 'unplugin'
import { createContext } from './ctx'

export default createUnplugin<Options>((options) => {
  let ctx = createContext(options)
  return {
    name: 'unplugin-auto-import',
    enforce: 'post',
    transformInclude(id) {
      return ctx.filter(id)
    },
    async transform(code, id) {
      return ctx.transform(code, id)
    },
    async buildStart() {
      await ctx.scanDirs()
    },
    async buildEnd() {
      await ctx.writeConfigFiles()
    },
    vite: {
      async config(config) {
        if (options.viteOptimizeDeps === false)
          return

        const exclude = config.optimizeDeps?.exclude || []

        const imports = new Set((await ctx.unimport.getImports()).map(i => i.from).filter(i => i.match(/^[a-z@]/) && !exclude.includes(i) && isPackageExists(i)))

        if (!imports.size)
          return

        return {
          optimizeDeps: {
            include: [...imports],
          },
        }
      },
      async handleHotUpdate({ file }) {
        const relativeFile = path.relative(ctx.root, slash(file))
        if (ctx.dirs?.some(dir => pm.isMatch(slash(relativeFile), slash(typeof dir === 'string' ? dir : dir.glob))))
          await ctx.scanDirs()
      },
      async configResolved(config) {
        if (ctx.root !== config.root) {
          ctx = createContext(options, config.root)
          await ctx.scanDirs()
        }
      },
    },
  }
})
