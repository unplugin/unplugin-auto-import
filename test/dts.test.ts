import { join } from 'node:path'
import process from 'node:process'
import { expect, it } from 'vitest'
import { createContext } from '../src/core/ctx'

describe('dts', () => {
  it('normal', async () => {
    const cwd = process.cwd()
    const ctx = createContext({
      packagePresets: ['magic-string'],
      imports: [
        'vue-demi',
        'react',
        'react-dom',
        'svelte',
        'svelte/animate',
        'svelte/easing',
        'svelte/motion',
        'svelte/store',
        'svelte/transition',
        {
          custom: [
            'customNamed',
            ['default', 'customDefault'],
          ],
          custom2: [
            ['*', 'custom2'],
          ],
          [join(cwd, 'foo.ts')]: ['foo'],
        },
        'vue/macros',
      ],
    })

    expect(await ctx.generateDTS(join(cwd, 'index.d.ts'))).toMatchSnapshot()
  })

  it('dts ignore', async () => {
    const cwd = process.cwd()
    const ctx = createContext({
      imports: [{
        custom: [
          'shouldBePresent',
          'shouldAlsoBePresent',
          'shouldBeIgnored',
          'ignoreme_shoudAlsoBeIgnored',
        ],
      }],
      ignoreDts: [
        'shouldBeIgnored',
        /^ignoreme_/,
      ],
    })

    const dtsContent = await ctx.generateDTS(join(cwd, './test/tmp/dts.ignore.d.ts'))

    expect(dtsContent).toContain('shouldBePresent')
    expect(dtsContent).toContain('shouldAlsoBePresent')
    expect(dtsContent).not.toContain('shouldBeIgnored')
    expect(dtsContent).not.toContain('ignoreme_shoudAlsoBeIgnored')
  })

  it('dts in append mode', async () => {
    const cwd = process.cwd()
    const dts = join(cwd, './test/tmp/dts.increase.d.ts')
    const ctx = createContext({
      imports: ['vue'],
      dts,
    })

    const dtsContent = await ctx.generateDTS(dts)
    expect(dtsContent).toContain('AAA')
    expect(dtsContent).toContain('BBB')
    expect(dtsContent).toContain('$$')
    expect(dtsContent).toContain('customFile')
    expect(dtsContent).toContain('customFile1')
  })

  it('dts in overwrite mode', async () => {
    const cwd = process.cwd()
    const dts = join(cwd, './test/tmp/dts.increase.d.ts')
    const ctx = createContext({
      imports: ['vue'],
      dts,
      dtsMode: 'overwrite',
    })

    const dtsContent = await ctx.generateDTS(dts)
    expect(dtsContent).not.toContain('AAA')
    expect(dtsContent).not.toContain('BBB')
    expect(dtsContent).not.toContain('$$')
    expect(dtsContent).not.toContain('customFile')
    expect(dtsContent).not.toContain('customFile1')

    expect(dtsContent).toContain('ref')
    expect(dtsContent).toContain('reactive')
    expect(dtsContent).toContain('watch')
    expect(dtsContent).toContain('computed')
  })
})
