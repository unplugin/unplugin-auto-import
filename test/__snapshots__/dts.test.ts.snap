// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dts 1`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// noinspection JSUnusedGlobalSymbols
// Generated by unplugin-auto-import
export {}
declare global {
  const $$: typeof import('vue/macros')['$$']
  const $: typeof import('vue/macros')['$']
  const $computed: typeof import('vue/macros')['$computed']
  const $customRef: typeof import('vue/macros')['$customRef']
  const $ref: typeof import('vue/macros')['$ref']
  const $shallowRef: typeof import('vue/macros')['$shallowRef']
  const $toRef: typeof import('vue/macros')['$toRef']
  const Bundle: typeof import('magic-string')['Bundle']
  const EffectScope: typeof import('vue-demi')['EffectScope']
  const SourceMap: typeof import('magic-string')['SourceMap']
  const afterUpdate: typeof import('svelte')['afterUpdate']
  const backIn: typeof import('svelte/easing')['backIn']
  const backInOut: typeof import('svelte/easing')['backInOut']
  const backOut: typeof import('svelte/easing')['backOut']
  const beforeUpdate: typeof import('svelte')['beforeUpdate']
  const blur: typeof import('svelte/transition')['blur']
  const bounceIn: typeof import('svelte/easing')['bounceIn']
  const bounceInOut: typeof import('svelte/easing')['bounceInOut']
  const bounceOut: typeof import('svelte/easing')['bounceOut']
  const circIn: typeof import('svelte/easing')['circIn']
  const circInOut: typeof import('svelte/easing')['circInOut']
  const circOut: typeof import('svelte/easing')['circOut']
  const computed: typeof import('vue-demi')['computed']
  const createApp: typeof import('vue-demi')['createApp']
  const createEventDispatcher: typeof import('svelte')['createEventDispatcher']
  const createRef: typeof import('react')['createRef']
  const crossfade: typeof import('svelte/transition')['crossfade']
  const cubicIn: typeof import('svelte/easing')['cubicIn']
  const cubicInOut: typeof import('svelte/easing')['cubicInOut']
  const cubicOut: typeof import('svelte/easing')['cubicOut']
  const custom2: typeof import('custom2')
  const customDefault: typeof import('custom')['default']
  const customNamed: typeof import('custom')['customNamed']
  const customRef: typeof import('vue-demi')['customRef']
  const default: typeof import('magic-string')['default']
  const defineAsyncComponent: typeof import('vue-demi')['defineAsyncComponent']
  const defineComponent: typeof import('vue-demi')['defineComponent']
  const derived: typeof import('svelte/store')['derived']
  const draw: typeof import('svelte/transition')['draw']
  const effectScope: typeof import('vue-demi')['effectScope']
  const elasticIn: typeof import('svelte/easing')['elasticIn']
  const elasticInOut: typeof import('svelte/easing')['elasticInOut']
  const elasticOut: typeof import('svelte/easing')['elasticOut']
  const expoIn: typeof import('svelte/easing')['expoIn']
  const expoInOut: typeof import('svelte/easing')['expoInOut']
  const expoOut: typeof import('svelte/easing')['expoOut']
  const fade: typeof import('svelte/transition')['fade']
  const flip: typeof import('svelte/animate')['flip']
  const fly: typeof import('svelte/transition')['fly']
  const foo: typeof import('./foo')['foo']
  const forwardRef: typeof import('react')['forwardRef']
  const get: typeof import('svelte/store')['get']
  const getAllContexts: typeof import('svelte')['getAllContexts']
  const getContext: typeof import('svelte')['getContext']
  const getCurrentInstance: typeof import('vue-demi')['getCurrentInstance']
  const getCurrentScope: typeof import('vue-demi')['getCurrentScope']
  const h: typeof import('vue-demi')['h']
  const hasContext: typeof import('svelte')['hasContext']
  const inject: typeof import('vue-demi')['inject']
  const isProxy: typeof import('vue-demi')['isProxy']
  const isReactive: typeof import('vue-demi')['isReactive']
  const isReadonly: typeof import('vue-demi')['isReadonly']
  const isRef: typeof import('vue-demi')['isRef']
  const lazy: typeof import('react')['lazy']
  const linear: typeof import('svelte/easing')['linear']
  const markRaw: typeof import('vue-demi')['markRaw']
  const memo: typeof import('react')['memo']
  const nextTick: typeof import('vue-demi')['nextTick']
  const onActivated: typeof import('vue-demi')['onActivated']
  const onBeforeMount: typeof import('vue-demi')['onBeforeMount']
  const onBeforeUnmount: typeof import('vue-demi')['onBeforeUnmount']
  const onBeforeUpdate: typeof import('vue-demi')['onBeforeUpdate']
  const onDeactivated: typeof import('vue-demi')['onDeactivated']
  const onDestroy: typeof import('svelte')['onDestroy']
  const onErrorCaptured: typeof import('vue-demi')['onErrorCaptured']
  const onMount: typeof import('svelte')['onMount']
  const onMounted: typeof import('vue-demi')['onMounted']
  const onScopeDispose: typeof import('vue-demi')['onScopeDispose']
  const onServerPrefetch: typeof import('vue-demi')['onServerPrefetch']
  const onUnmounted: typeof import('vue-demi')['onUnmounted']
  const onUpdated: typeof import('vue-demi')['onUpdated']
  const provide: typeof import('vue-demi')['provide']
  const quadIn: typeof import('svelte/easing')['quadIn']
  const quadInOut: typeof import('svelte/easing')['quadInOut']
  const quadOut: typeof import('svelte/easing')['quadOut']
  const quartIn: typeof import('svelte/easing')['quartIn']
  const quartInOut: typeof import('svelte/easing')['quartInOut']
  const quartOut: typeof import('svelte/easing')['quartOut']
  const quintIn: typeof import('svelte/easing')['quintIn']
  const quintInOut: typeof import('svelte/easing')['quintInOut']
  const quintOut: typeof import('svelte/easing')['quintOut']
  const reactive: typeof import('vue-demi')['reactive']
  const readable: typeof import('svelte/store')['readable']
  const readonly: typeof import('vue-demi')['readonly']
  const ref: typeof import('vue-demi')['ref']
  const scale: typeof import('svelte/transition')['scale']
  const setContext: typeof import('svelte')['setContext']
  const shallowReactive: typeof import('vue-demi')['shallowReactive']
  const shallowReadonly: typeof import('vue-demi')['shallowReadonly']
  const shallowRef: typeof import('vue-demi')['shallowRef']
  const sineIn: typeof import('svelte/easing')['sineIn']
  const sineInOut: typeof import('svelte/easing')['sineInOut']
  const sineOut: typeof import('svelte/easing')['sineOut']
  const slide: typeof import('svelte/transition')['slide']
  const spring: typeof import('svelte/motion')['spring']
  const startTransition: typeof import('react')['startTransition']
  const tick: typeof import('svelte')['tick']
  const toRaw: typeof import('vue-demi')['toRaw']
  const toRef: typeof import('vue-demi')['toRef']
  const toRefs: typeof import('vue-demi')['toRefs']
  const toValue: typeof import('vue-demi')['toValue']
  const triggerRef: typeof import('vue-demi')['triggerRef']
  const tweened: typeof import('svelte/motion')['tweened']
  const unref: typeof import('vue-demi')['unref']
  const useAttrs: typeof import('vue-demi')['useAttrs']
  const useCallback: typeof import('react')['useCallback']
  const useContext: typeof import('react')['useContext']
  const useCssModule: typeof import('vue-demi')['useCssModule']
  const useDebugValue: typeof import('react')['useDebugValue']
  const useDeferredValue: typeof import('react')['useDeferredValue']
  const useEffect: typeof import('react')['useEffect']
  const useId: typeof import('react')['useId']
  const useImperativeHandle: typeof import('react')['useImperativeHandle']
  const useInsertionEffect: typeof import('react')['useInsertionEffect']
  const useLayoutEffect: typeof import('react')['useLayoutEffect']
  const useMemo: typeof import('react')['useMemo']
  const useReducer: typeof import('react')['useReducer']
  const useRef: typeof import('react')['useRef']
  const useSlots: typeof import('vue-demi')['useSlots']
  const useState: typeof import('react')['useState']
  const useSyncExternalStore: typeof import('react')['useSyncExternalStore']
  const useTransition: typeof import('react')['useTransition']
  const watch: typeof import('vue-demi')['watch']
  const watchEffect: typeof import('vue-demi')['watchEffect']
  const watchPostEffect: typeof import('vue-demi')['watchPostEffect']
  const watchSyncEffect: typeof import('vue-demi')['watchSyncEffect']
  const writable: typeof import('svelte/store')['writable']
}
// for type re-export
declare global {
  // @ts-ignore
  export type { Component, Slot, Slots, ComponentPublicInstance, ComputedRef, DirectiveBinding, ExtractDefaultPropTypes, ExtractPropTypes, ExtractPublicPropTypes, InjectionKey, PropType, Ref, MaybeRef, MaybeRefOrGetter, VNode, WritableComputedRef } from 'vue-demi'
  import('vue-demi')
}
"
`;
